### PR TITLE
[dvsim] Separate publish report from dvsim flow [PART1]

### PR DESCRIPTION
**Pull request by cindychip**
This PR is the first part of moving publish report to a separate script.
This PR moves the generated results into a separate /scratch_root/report
folder.
The result summary will be located at: /scratch_root/{branch}/report
The individual summary will be located at: /scratch_path/report

I also update the --purge option so it won't delete the report folders.

For output results, this PR modifies the dvsim to output:
1. results_{date-time}.html
2. publish_results_{date-time}.html (different from the previous file, this one has a URL link to all previous published results)
3. summary_{data-time}.html 

Signed-off-by: Cindy Chen <chencindy@opentitan.org>

### CODE REVIEW

File: hw/data/common_project_cfg.hjson [hunk 1]
```diff
@@ -8,8 +8,10 @@
   results_server:   reports.opentitan.org
 
   // Default directory structure for the output
-  scratch_base_path:  "{scratch_root}/{branch}"
-  scratch_path:       "{scratch_base_path}/{dut}-{flow}-{tool}"
+  scratch_base_path:   "{scratch_root}/{branch}"
+  scratch_path:        "{scratch_base_path}/{dut}-{flow}-{tool}"
+  report_summary_path: "{scratch_root}/{branch}/report/summary"
+  report_path:         "{scratch_root}/{branch}/report/{dut}-{flow}-{tool}"
```
Comment (sriyerg) [inline hw/data/common_project_cfg.hjson:14]: Hmm, now that I think about it more, it seems like we could simply reuse the `scratch_path` to find the HTML report as well. Thoughts?

File: util/dvsim/SimCfg.py [hunk 2]
```diff
@@ -672,8 +673,8 @@ def create_bucket_report(buckets):
 
         self.results_md = results_str
 
-        # Write results to the scratch area
-        results_path = self.scratch_path + "/results_" + self.timestamp + ".md"
+        # Write results to the scratch report area
```
Comment (sriyerg) [inline util/dvsim/SimCfg.py:676]: I dont think anyone has ever looked at this md file. We should simply create the HTML file instead (which we will publish to the report server).

File: hw/data/common_project_cfg.hjson (hunk 1 above)
Comment (cindychip) [reply hw/data/common_project_cfg.hjson:14]: true, then we can avoid adding another variable. I just add it here cause then I can use the `.mk` file to create each folder for reports (I mean the folder: `{scratch_root}/{branch}/report/{dut}-{flow}-{tool}`)
I do not have any strong preference, if using `scratch_path`, we can avoid this variable, and use a for loop to create each folder in `dvsim` :)

File: util/dvsim/SimCfg.py (hunk 2 above)
Comment (cindychip) [reply util/dvsim/SimCfg.py:676]: sure, I am fine with deleting this `.md` file.

File: hw/data/common_project_cfg.hjson (hunk 1 above)
Comment (sriyerg) [reply hw/data/common_project_cfg.hjson:14]: Yes, that's right. Use a fixed directory called `report` under `scratch_path` to put the generated HTML and all other report files.

File: util/dvsim/FlowCfg.py [hunk 3]
```diff
@@ -582,24 +605,23 @@ def publish_results(self):
...
     def publish_results_summary(self):
         '''Public facing API for publishing md format results to the opentitan
         web server.
         '''
-        results_html_file = "summary_" + self.timestamp + ".html"
         results_page_url = self.results_summary_server_page.replace(
             self.results_server_prefix, self.results_server_url_prefix)
 
         # Publish the results page.
-        # First, write the results html file temporarily to the scratch area.
```
Comment (cindychip) [inline util/dvsim/FlowCfg.py:594]: the results summary html file is created and output in line 410 of this file.

File: util/dvsim/FlowCfg.py [hunk 4]
```diff
@@ -393,20 +396,41 @@ def gen_results(self, results):
...
             log.info("[results]: [%s]:\n%s\n", item.name, result)
             log.info("[scratch_path]: [%s] [%s]", item.name, item.scratch_path)
+            log.info("[report_path]: [%s] [%s]", item.name, item.report_path)
             self.errors_seen |= item.errors_seen
 
+        self.create_report_dir(self.report_summary_path)
         if self.is_primary_cfg:
             self.gen_results_summary()
+            self.write_results_summary_html()
         self.gen_email_html_summary()
```
Comment (cindychip) [inline util/dvsim/FlowCfg.py:411]: @sriyerg I think for next PR, we can remove the email_html_summary.
We can output all summaries in html and let user decide which one to use for emailing.

To accomplish this, I think we can create the latest output files without the time/date postfix, and use the script to rename previous output. So user will know which file has the latest output.

File: util/dvsim/FlowCfg.py [hunk 5]
```diff
@@ -550,7 +574,7 @@ def _publish_results(self):
 
         # Publish the results page.
         # First, write the results html file temporarily to the scratch area.
-        results_html_file = self.scratch_path + "/results_" + self.timestamp + \
+        results_html_file = self.report_path + "/publish_results_" + self.timestamp + \
```
Comment (cindychip) [inline util/dvsim/FlowCfg.py:577]: add prefix `publish_` to distinguish with `results_{time-date}.html`.
This publish html file also contains a URL list of previous published path.

File: util/dvsim/FlowCfg.py [hunk 6]
```diff
@@ -393,20 +396,41 @@ def gen_results(self, results):
...
     def gen_results_summary(self):
         '''Public facing API to generate summary results for each IP/cfg file
         '''
         return
 
+    def write_results_html(self):
+        # Write results to the report area
+        results_html = md_results_to_html(self.results_title, self.css_file,
+                                          self.results_md)
+        result_path = self.report_path + "/results_" + self.timestamp + ".html"
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:422]:
```suggestion
        result_path = self.report_path + "/results.html"
```

File: util/dvsim/FlowCfg.py [hunk 7]
```diff
@@ -393,20 +396,41 @@ def gen_results(self, results):
...
+        with open(result_path, "w") as results_file:
+            results_file.write(results_html)
+        log.log(VERBOSE, "[results page]: [%s][%s], self.name, results_path")
+
+    def write_results_summary_html(self):
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:427]: Combine with prev function, use `is_parimary_cfg` flag.

File: util/dvsim/FlowCfg.py [hunk 8]
```diff
@@ -393,20 +396,41 @@ def gen_results(self, results):
 
         '''
         for item in self.cfgs:
+            item.report_path = os.path.join(item.scratch_path, "report")
+            item.create_report_dir(item.report_path)
             result = item._gen_results(results)
             log.info("[results]: [%s]:\n%s\n", item.name, result)
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:402]: Call  `       item.write_results_html()` here.

File: util/dvsim/OneShotCfg.py [hunk 9]
```diff
@@ -112,8 +112,11 @@ def _expand(self):
     # Purge the output directories. This operates on self.
     def _purge(self):
         assert self.scratch_path
-        log.info("Purging scratch path %s", self.scratch_path)
-        rm_path(self.scratch_path)
+        log.info("Purging scratch path %s except report folders", self.scratch_path)
```
Comment (sriyerg) [inline util/dvsim/OneShotCfg.py:115]: Revert.

File: util/dvsim/SimCfg.py [hunk 10]
```diff
@@ -671,14 +674,7 @@ def create_bucket_report(buckets):
...
         self.results_md = results_str
-
-        # Write results to the scratch area
-        results_path = self.scratch_path + "/results_" + self.timestamp + ".md"
-        with open(results_path, 'w') as results_file:
-            results_file.write(self.results_md)
-
-        # Return only the tables
-        log.log(VERBOSE, "[results page]: [%s] [%s]", self.name, results_path)
+        self.write_results_html()
```
Comment (sriyerg) [inline util/dvsim/SimCfg.py:677]: Move to FlowCfg.

**Review by cindychip — COMMENTED** [review]
Thanks for all the reviews! Updated them as you recommended.

**Review by sriyerg — APPROVED** [review]
LGTM - lets progressively make the logic tighter.

**Pull request merged**